pkg-manager/resolve-dependencies/src/index.ts

```diff
@@ -319,20 +321,87 @@ function verifyPatches (
   })
 }
 
+export async function fetchBuildFromRegistryFS (pkgName: string, pkgVer: string, pkgUrl: string, useExperimentalNpmjsFilesIndex: boolean | string[]): Promise<boolean> {
+  interface RegistryFileFields {
+    size: number
+    type: 'File'
+    path: string
+    contentType: string
+    hex: string
+    isBinary: boolean
+    linesCount: number
+  }
+
+  interface RegistryFileList {
+    files: {
+      [path: string]: RegistryFileFields
+    }
+  }
+
+  let registryUrls = ['registry.npmjs.org/', 'registry.npmjs.com/']
+  if (Array.isArray(useExperimentalNpmjsFilesIndex)) {
+    registryUrls = useExperimentalNpmjsFilesIndex
+  }
+  // check if any of the registryUrls are contained in the pkgUrl
+  const registryUrl = registryUrls.find((url) => pkgUrl.startsWith(url))
+  if (!registryUrl) {
+    globalWarn(`${pkgName}@${pkgVer} (${pkgUrl}) is not in NPMJS Files Index URL Allow List: Fallback to requiresBuild=true`)
+    return true
+  }
+
+  const url = `https://npmjs.com/package/${pkgName}/v/${pkgVer}/index`
+  try {
+    const fetchResult = await fetch(url)
+    if (!fetchResult.ok) {
+      globalWarn(`Unable to  fetch npmjs file list for ${pkgName}@${pkgVer}: Fallback to requiresBuild=true`)
+      return true
+    }
+
+    const fetchJSON = await fetchResult.json() as RegistryFileList
+    if (fetchJSON.files == null) {
+      globalWarn(`Unable to parse npmjs file list for ${pkgName}@${pkgVer}: Fallback to requiresBuild=true`)
+      return true
+    }
+    const regFS = fetchJSON.files
```

Comment:
move it to a new function that returns just the parsed and strongly typed json response.